### PR TITLE
feat: Anvil themeing and Anvil vertical alignment

### LINKED ISSUE #28591: [EPIC]: Theming
Based on the new [theme-ing mechanism in WDS](https://www.notion.so/appsmith/Theming-4aceae7fff4c4e2788e11a7a3e89d6ae?pvs=4), certain changes needs to be made to the Anvil layout system, such that the themeing is applied cohesively to the layout instead of just the widgets.

This Epic makes the necessary changes.

Themeing token reference: https://wds.appsmith.com/?path=/docs/design-system-theme-theme--docs

### LINKED ISSUE #28592: [Task]: Integrate Color tokens for backgrounds and other UI

### LINKED ISSUE #28593: [Task]: Use spacing tokens in layouts
Replace static values from padding, rowGap and columnGap with a spacing token.

Current decision is to use ``spacing-3`` token.

### LINKED ISSUE #29073: [EPIC]: Widget Vertical alignment
- Widgets by default are not aligned properly. This causes an asymmetry between the widgets and effects the visual aesthetics of an app, and reduces the quality that users can achieve with little work.

- Goal:
  - The following layout is aligned properly by default

Screenshot 2023-11-24 at 11 59 53 AM
  
  
- First hypothesis for a solution
https://www.notion.so/appsmith/Vertical-alignment-of-small-widgets-4ddeeb34ee874aab8685df67203d1fb7?pvs=4

### PR DESCRIPTION
**Pull request by riodeuno**
## Description
This PR adds the features of proper vertical alignment and themeing to Anvil.
- A separate `Container` component is created for Anvil, that is used as the layer on top of which the themeing tokens are applied.
- A default `min-height` is set using tokens for all widgets in Anvil. 
- Anvil now stops considering any `min-height` configurations provided by the widgets. It is the widgets responsibility to take care of their own heights, and Anvil will accommodate them -- no matter the height.
- Table widget's default height is now set to the min height that was configured for it earlier.
- `AnvilFlexComponent` now has `overflow:visible` to allow the shadows for zones and sections to not be cut-off.
- All widgets are aligned center vertically by default. This will apply if they're smaller than the set `min-height`
- Zones and Sections have elevation styles applied suing the `Container` component mentioned above.
- Zones and Sections don't have any styling property other than `Background`, we'll add more as we understand more about the product.
   

> Conditional vertical margin applied to widgets.
>     If in a row of widgets (.aligned-widget-row), one of the widgets has a label ([data-field-label-wrapper]), then
>     all widgets (.anvil-widget-wrapper) in the row other than the widget with the label, will shift down using the 
>     margin-block-start property. This is to ensure that the widgets are aligned vertically.
>     The value of the margin-block-start property is calculated based on the spacing tokens used by the labels in input 
>     like components
> 

#### PR fixes following issue(s)
Fixes #29073 
Fixes #28591
Fixes #28592 
Fixes #28593

#### Media
Screenshot 2023-12-28 at 3 55 05 AM

#### Type of change
- New feature (non-breaking change which adds functionality)
#### How Has This Been Tested?
- [x] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress

#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
	- Introduced conditional vertical margins for widgets to ensure alignment within rows.
	- Added a new `Container` component for thematic elevation styles in Anvil widgets.
	- Implemented elevation style options and semantic background settings for Section and Zone widgets.

- **Enhancements**
	- Improved visual layout and alignment of AnvilFlexComponent with updated styling properties.
	- Added `className` properties to various layout components for enhanced CSS targeting.

- **Style**
	- Updated widget styles to accommodate new background and elevation features.

- **Refactor**
	- Simplified padding logic in WDSParagraphWidget.
	- Streamlined dimensions calculation in WDSTableWidget.

- **Documentation**
	- Renamed sections in property panes to better reflect background styling options.

- **Chores**
	- Added `Elevations` enum to manage elevation values consistently across components.

### CODE REVIEW

File: app/client/src/index.css [hunk 1]
```diff
@@ -23,6 +23,11 @@ body {
   color: var(--ads-v2-color-fg);
 }
 
+.aligned-widget-row:has(.anvil-widget-wrapper [data-field-label-wrapper])
+  .anvil-widget-wrapper:not(:has([data-field-label-wrapper])) {
+  margin-top: calc(var(--inner-spacing-2) + var(--sizing-3));
```
Comment (riodeuno) [inline app/client/src/index.css:28]: @riodeuno use `margin-block-start`

@KelvinOm This value is the amount of space the label in an input widget takes. For vertical alignment (with or without lables), we're applying this margin. Could this change in the future? If so, can we create a new variable `--label-size`, that we can reuse in input widgets as well as here?

File: app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx [hunk 2]
```diff
@@ -73,9 +73,6 @@ class WDSParagraphWidget extends BaseWidget<TextWidgetProps, WidgetState> {
         isBold={this.props?.fontStyle?.includes("bold")}
         isItalic={this.props?.fontStyle?.includes("italic")}
         lineClamp={this.props.lineClamp ? this.props.lineClamp : undefined}
-        style={{
```
Comment (riodeuno) [inline app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx:76]: @jsartisan This `paddingBottom` is preventing the text widget from being perfectly vertically aligned. What is my alternative here? Also, why did we add this?

File: app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx [hunk 3]
```diff
@@ -1182,15 +1182,15 @@ export class WDSTableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
...
+    // let { componentHeight } = this.props;
     // Hacky fix for now to supply width to table widget
     let componentWidth: number =
       document
         .getElementById(getAnvilWidgetDOMId(this.props.widgetId))
         ?.getBoundingClientRect().width || this.props.componentWidth;
     // (2 * WIDGET_PADDING) gives the total horizontal padding (i.e. paddingLeft + paddingRight)
     componentWidth = componentWidth - 2 * WIDGET_PADDING;
-    return { componentHeight, componentWidth };
+    return { componentHeight: 300, componentWidth };
```
Comment (riodeuno) [inline app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx:1193]: @KelvinOm I'm changing the use of `minHeight` in Anvil. Anvil will always use the component height. As a result I've also hardcoded the `componentHeight` to `300` -- which was the minimum height before. I'm assuming we're going to re-visit this, so there is no harm with this change?

File: app/client/src/ce/entities/FeatureFlag.ts [hunk 4]
```diff
@@ -59,7 +59,7 @@ export const DEFAULT_FEATURE_FLAG_VALUE: FeatureFlags = {
   release_appnavigationlogoupload_enabled: false,
   release_embed_hide_share_settings_enabled: false,
   ab_gsheet_schema_enabled: false,
-  ab_wds_enabled: false,
+  ab_wds_enabled: true,
```
Comment (riodeuno) [inline app/client/src/ce/entities/FeatureFlag.ts:62]: @riodeuno This PR is ready, let's revert this change.

File: app/client/src/ce/sagas/userSagas.tsx [hunk 5]
```diff
@@ -541,6 +541,8 @@ export function* fetchFeatureFlags() {
     );
     const isValidResponse: boolean = yield validateResponse(response);
     if (isValidResponse) {
+      response.data.release_anvil_enabled = true;
```
Comment (riodeuno) [inline app/client/src/ce/sagas/userSagas.tsx:544]: @riodeuno This PR is ready, let's revert this change

File: app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx [hunk 6]
```diff
@@ -76,7 +76,7 @@ export function AnvilFlexComponent(props: AnvilFlexComponentProps) {
   );
 
   const stopEventPropagation = (e: MouseEvent<HTMLElement>) => {
-    !isSnipingMode && e.stopPropagation();
+    !isSnipingMode && e.preventDefault();
```
Comment (riodeuno) [inline app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx:79]: @riodeuno Revert this change as this has been handled in another PR

File: app/client/src/index.css (hunk 1 above)
Comment (KelvinOm) [reply app/client/src/index.css:28]: @riodeuno I don't like this place either. I don't know yet how to count this correctly. I want to make this a separate task, as I probably will need to make changes to the components themselves. Now I see a issue with the way we set the height in our components. Now it's a combination of padding and height values. I'm thinking about just set the height with 1 token.

File: app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx (hunk 2 above)
Comment (KelvinOm) [reply app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx:76]: We've been correcting this behavior here (see the picture). I think it's worth checking how it works with your changes, maybe it's no longer relevant. But it's worth checking how the widget works if the canvas is more than 800px.

292200899-14674779-f04a-4f81-91a3-78430499f4b8

File: app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx (hunk 3 above)
Comment (KelvinOm) [reply app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx:1193]: @riodeuno Yes, I will still work with the table. It's okay now.

File: app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx [hunk 7]
```diff
@@ -1182,15 +1182,15 @@ export class WDSTableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
 
   getPaddingAdjustedDimensions = () => {
     // eslint-disable-next-line prefer-const
-    let { componentHeight } = this.props;
+    // let { componentHeight } = this.props;
```
Comment (KelvinOm) [inline app/client/src/widgets/wds/WDSTableWidget/widget/index.tsx:1185]: Don't forget about this. Seems like no more relevant.

File: app/client/src/widgets/anvil/Container.tsx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+  justifyContent?: string;
+  alignItems?: string;
+  dropDisabled?: boolean;
+  layoutSystemType?: LayoutSystemTypes;
+  isListItemContainer?: boolean;
+  elevation: Elevations;
+  elevatedBackground: boolean;
+}
+
+export default ContainerComponent;
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:207]: Let's use a named import

File: app/client/src/widgets/anvil/Container.tsx [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+      resizeDisabled={props.resizeDisabled}
+      type={props.type}
+      widgetId={props.widgetId}
+    >
+      {props.children}
+    </ContainerComponentWrapper>
+  );
+}
+
+export type ContainerStyle = "border" | "card" | "rounded-border" | "none";
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:182]: It doesn't seem to be used anywhere

File: app/client/src/widgets/anvil/Container.tsx [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+export interface ContainerComponentProps extends WidgetStyleContainerProps {
+  children?: ReactNode;
+  shouldScrollContents?: boolean;
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:186]: Why we need this property? I don't see it being used anywhere

File: app/client/src/widgets/anvil/Container.tsx [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+  resizeDisabled?: boolean;
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:187]: Is it for redistribution behavior?

File: app/client/src/widgets/anvil/Container.tsx [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+          if (!isMainContainer) {
+            el.classList.add("hover-styles");
+          }
+        });
+      }
+    },
+    [props.onClickCapture],
+  );
+
+  // console.log("####", props);
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:117]: Don't forget to delete this.

File: app/client/src/widgets/anvil/Container.tsx [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+export type ContainerStyle = "border" | "card" | "rounded-border" | "none";
+
+export interface ContainerComponentProps extends WidgetStyleContainerProps {
+  children?: ReactNode;
+  shouldScrollContents?: boolean;
+  resizeDisabled?: boolean;
+  selected?: boolean;
+  focused?: boolean;
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:189]: I don't see where selected and focused properties are used. What is the difference between it in the context of zones and sections? As I understand it, they can't have a focus state.

File: app/client/src/widgets/anvil/Container.tsx [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+>`
+  height: 100%;
+  width: 100%;
+  overflow: hidden;
+  outline: none;
+  border: none;
+  ${(props) => (!!props.dropDisabled ? `position: relative;` : ``)}
+
+  ${(props) =>
+    props.elevatedBackground
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:30]: I would not do this logic at all yet. I think it's better to leave only zones behavior. [Taras wrote about it yesterday](https://theappsmith.slack.com/archives/C025SE88KNE/p1703692631144889). A more complex logic should be implemented here.

cc @tbrizitsky

File: app/client/src/widgets/anvil/Container.tsx [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+          : ""
+      }`}
+      dropDisabled={props.dropDisabled}
+      elevatedBackground={props.elevatedBackground}
+      elevation={props.elevation}
+      onClick={props.onClick}
+      onClickCapture={props.onClickCapture}
+      onMouseOver={onMouseOver}
+      ref={containerRef}
+      resizeDisabled={props.resizeDisabled}
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:139]: seems like resizeDisabled and type props are not needed for styling

File: app/client/src/widgets/anvil/Container.tsx [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+      tabIndex={0}
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:140]: Why do we need `tabIndex` here? As I understand this component is not designed to work with the keyboard.

File: app/client/src/widgets/anvil/Container.tsx [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+  type: WidgetType;
+  dropDisabled?: boolean;
+  elevation: Elevations;
+  elevatedBackground: boolean;
+}
+function ContainerComponentWrapper(
+  props: PropsWithChildren<ContainerWrapperProps>,
+) {
+  const containerRef: RefObject<HTMLDivElement> = useRef<HTMLDivElement>(null);
+  const layoutSystemType = useSelector(getLayoutSystemType);
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:60]: Why do we need to determine the type of layout system here? As I understand everything should work only with anvil.

File: app/client/src/widgets/anvil/Container.tsx [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+
+  useEffect(() => {
+    const supportsNativeSmoothScroll =
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:63]: Could you explain why this is need?

File: app/client/src/widgets/anvil/Container.tsx [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+      }
+    },
+    [props.onClickCapture],
+  );
+
+  // console.log("####", props);
+
+  return (
+    <StyledContainerComponent
+      // Before you remove: generateClassName is used for bounding the resizables within this canvas
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:121]: Seems like this comment not relevant anymore.

File: app/client/src/widgets/anvil/Container.tsx [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+      // getCanvasClassName is used to add a scrollable parent.
+      backgroundColor={props.backgroundColor}
+      className={`${generateClassName(
```
Comment (KelvinOm) [inline app/client/src/widgets/anvil/Container.tsx:124]: The classNames are also probably left over from the old implementation, right?

File: app/client/src/index.css (hunk 1 above)
Comment (riodeuno) [reply app/client/src/index.css:28]: @KelvinOm Got it. Let's make this a separate task. I'll update the code once we tackle it properly. I'll keep this as it is for now.

File: app/client/src/widgets/anvil/Container.tsx (hunk 14 above)
Comment (riodeuno) [reply app/client/src/widgets/anvil/Container.tsx:30]: @KelvinOm @tbrizitsky 's logic is a hypothesis at the moment. I'll tackle it separately. I would like to follow through with the original hypothesis for now, and update the logic once we have a new or more optimised approach.

The product updates in this PR, is something I would not want to change at this point, in this PR.

File: app/client/src/widgets/anvil/Container.tsx (hunk 17 above)
Comment (riodeuno) [reply app/client/src/widgets/anvil/Container.tsx:60]: Yes, I'm the process of cleaning up this component.

File: app/client/src/widgets/anvil/Container.tsx (hunk 18 above)
Comment (riodeuno) [reply app/client/src/widgets/anvil/Container.tsx:63]: This is needed for context switching. It was a part of the original implementation. I'll revisit this.

File: app/client/src/widgets/anvil/Container.tsx (hunk 20 above)
Comment (riodeuno) [reply app/client/src/widgets/anvil/Container.tsx:124]: Yes. I'll update this.

File: app/client/src/widgets/anvil/Container.tsx (hunk 15 above)
Comment (riodeuno) [reply app/client/src/widgets/anvil/Container.tsx:139]: Yes, they're no longer necessary.

File: app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx (hunk 2 above)
Comment (riodeuno) [reply app/client/src/widgets/wds/WDSParagraphWidget/widget/index.tsx:76]: Seems like it's not an issue anymore.

Screenshot 2023-12-28 at 10 18 09 PM

File: app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx [hunk 21]
```diff
@@ -158,7 +158,7 @@ export function AnvilFlexComponent(props: AnvilFlexComponentProps) {
     return {
       position: "relative",
       // overflow is set to make sure widgets internal components/divs don't overflow this boundary causing scrolls
-      overflow: "hidden",
+      overflow: "visible",
```
Comment (prsidhu) [inline app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx:161]: @riodeuno what is the purpose of this?

Comment (riodeuno) [reply app/client/src/layoutSystems/anvil/common/AnvilFlexComponent.tsx:161]: @prsidhu 
> `AnvilFlexComponent` now has `overflow:visible` to allow the shadows for zones and sections to not be cut-off.

**Review by prsidhu — APPROVED** [review]

**Pull request merged**